### PR TITLE
orchestra/remote.py: raise if ensure_online() fails

### PR DESCRIPTION
**Pull request by tchaikov**

### CODE REVIEW

**Comment by tchaikov** [comment]
see also http://pulpito.ceph.com/kchai-2020-03-25_08:14:01-rados-wip-kefu-testing-2020-03-25-1220-distro-basic-smithi/4888710/

**Comment by kshtsk** [comment]
@susebot run deploy

**Comment by kshtsk** [comment]
@vasukulkarni @tchaikov this change introduced a regression for us

File: teuthology/orchestra/remote.py [hunk 1]
```diff
@@ -191,10 +196,7 @@ def run(self, **kwargs):
 
         TODO refactor to move run.run here?
         """
-        if not self.ssh or \
-           not self.ssh.get_transport() or \
-           not self.ssh.get_transport().is_active():
-            self.reconnect()
```
Comment (kshtsk) [inline teuthology/orchestra/remote.py:197]: @tchaikov why have you remove `self.reconnect()` it was doing very useful thing when rebooting a node

Comment (kshtsk) [reply teuthology/orchestra/remote.py:197]: the `ensure_online()` does not do this.

**Comment by kshtsk** [comment]
Your issue was like:
```
2020-03-25T09:00:47.044 INFO:teuthology.orchestra.remote:Trying to reconnect to host
2020-03-25T09:00:47.045 DEBUG:teuthology.orchestra.connection:{'username': 'ubuntu', 'hostname': 'smithi088.front.sepia.ceph.com', 'timeout': 60}
2020-03-25T09:00:47.046 DEBUG:teuthology.orchestra.remote:[Errno None] Unable to connect to port 22 on 172.21.15.88
2020-03-25T09:00:47.046 ERROR:teuthology.run_tasks:Saw exception from tasks.
Traceback (most recent call last):
  File "/home/teuthworker/src/git.ceph.com_git_teuthology_master/teuthology/run_tasks.py", line 89, in run_tasks
    manager.__enter__()
  File "/usr/lib/python2.7/contextlib.py", line 17, in __enter__
    return self.gen.next()
  File "/home/teuthworker/src/git.ceph.com_git_teuthology_master/teuthology/task/internal/lock_machines.py", line 78, in lock_machines
    os_version, arch)
  File "/home/teuthworker/src/git.ceph.com_git_teuthology_master/teuthology/lock/ops.py", line 147, in lock_many
    update_nodes(reimaged)
  File "/home/teuthworker/src/git.ceph.com_git_teuthology_master/teuthology/lock/ops.py", line 33, in update_nodes
    inventory_info = remote.inventory_info
  File "/home/teuthworker/src/git.ceph.com_git_teuthology_master/teuthology/orchestra/remote.py", line 475, in inventory_info
    node['arch'] = self.arch
  File "/home/teuthworker/src/git.ceph.com_git_teuthology_master/teuthology/orchestra/remote.py", line 459, in arch
    self._arch = self.sh('uname -m').strip()
  File "/home/teuthworker/src/git.ceph.com_git_teuthology_master/teuthology/orchestra/remote.py", line 242, in sh
    proc=self.run(**kwargs)
  File "/home/teuthworker/src/git.ceph.com_git_teuthology_master/teuthology/orchestra/remote.py", line 198, in run
    r = self._runner(client=self.ssh, name=self.shortname, **kwargs)
  File "/home/teuthworker/src/git.ceph.com_git_teuthology_master/teuthology/orchestra/run.py", line 412, in run
    transport = client.get_transport()
AttributeError: 'NoneType' object has no attribute 'get_transport'
```
Please explain how is this change relevant to the stack trace above?

**Comment by kshtsk** [comment]
The reason of the failure was that after an unsuccessful try to reconnect:
```2020-03-25T09:00:47.046 DEBUG:teuthology.orchestra.remote:[Errno None] Unable to connect to port 22 on 172.21.15.88
```
There was no checked the return code of a reconnect call `.reconnect()`

File: teuthology/orchestra/remote.py (hunk 1 above)
Comment (tchaikov) [reply teuthology/orchestra/remote.py:197]: > it was doing very useful thing when rebooting a node

could you define "very useful thing"?

**Comment by tchaikov** [comment]
> Your issue was like:
> 
> ```
> 2020-03-25T09:00:47.044 INFO:teuthology.orchestra.remote:Trying to reconnect to host
> 2020-03-25T09:00:47.045 DEBUG:teuthology.orchestra.connection:{'username': 'ubuntu', 'hostname': 'smithi088.front.sepia.ceph.com', 'timeout': 60}
> 2020-03-25T09:00:47.046 DEBUG:teuthology.orchestra.remote:[Errno None] Unable to connect to port 22 on 172.21.15.88
> 2020-03-25T09:00:47.046 ERROR:teuthology.run_tasks:Saw exception from tasks.
> Traceback (most recent call last):
>   File "/home/teuthworker/src/git.ceph.com_git_teuthology_master/teuthology/run_tasks.py", line 89, in run_tasks
>     manager.__enter__()
>   File "/usr/lib/python2.7/contextlib.py", line 17, in __enter__
>     return self.gen.next()
>   File "/home/teuthworker/src/git.ceph.com_git_teuthology_master/teuthology/task/internal/lock_machines.py", line 78, in lock_machines
>     os_version, arch)
>   File "/home/teuthworker/src/git.ceph.com_git_teuthology_master/teuthology/lock/ops.py", line 147, in lock_many
>     update_nodes(reimaged)
>   File "/home/teuthworker/src/git.ceph.com_git_teuthology_master/teuthology/lock/ops.py", line 33, in update_nodes
>     inventory_info = remote.inventory_info
>   File "/home/teuthworker/src/git.ceph.com_git_teuthology_master/teuthology/orchestra/remote.py", line 475, in inventory_info
>     node['arch'] = self.arch
>   File "/home/teuthworker/src/git.ceph.com_git_teuthology_master/teuthology/orchestra/remote.py", line 459, in arch
>     self._arch = self.sh('uname -m').strip()
>   File "/home/teuthworker/src/git.ceph.com_git_teuthology_master/teuthology/orchestra/remote.py", line 242, in sh
>     proc=self.run(**kwargs)
>   File "/home/teuthworker/src/git.ceph.com_git_teuthology_master/teuthology/orchestra/remote.py", line 198, in run
>     r = self._runner(client=self.ssh, name=self.shortname, **kwargs)
>   File "/home/teuthworker/src/git.ceph.com_git_teuthology_master/teuthology/orchestra/run.py", line 412, in run
>     transport = client.get_transport()
> AttributeError: 'NoneType' object has no attribute 'get_transport'
> ```
> 
> Please explain how is this change relevant to the stack trace above?

i hope the commit message could help:

> for better readability, and to have better errors when we fail to connect to remote.

**Comment by tchaikov** [comment]
> The reason of the failure was that after an unsuccessful try to reconnect:
> 
> ```
> 2020-03-25T09:00:47.046 DEBUG:teuthology.orchestra.remote:[Errno None] Unable to connect to port 22 on 172.21.15.88
> ```
> 
> There was no checked the return code of a reconnect call `.reconnect()`

thanks for the insights. but what is your intention here?

**Comment by kshtsk** [comment]
> > The reason of the failure was that after an unsuccessful try to reconnect:
> > ```
> > 2020-03-25T09:00:47.046 DEBUG:teuthology.orchestra.remote:[Errno None] Unable to connect to port 22 on 172.21.15.88
> > ```
> > 
> > 
> > There was no checked the return code of a reconnect call `.reconnect()`
> 
> thanks for the insights. but what is your intention here?

This change introduced a regression https://tracker.ceph.com/issues/45255

File: teuthology/orchestra/remote.py (hunk 1 above)
Comment (kshtsk) [reply teuthology/orchestra/remote.py:197]: the name of function "reconnect" is self explanatory, I suppose

Comment (tchaikov) [reply teuthology/orchestra/remote.py:197]: it does. "reconnect" connect without passing create_key or host_key.

**Comment by tchaikov** [comment]
> @tchaikov This PR introduced a regression https://tracker.ceph.com/issues/45255

ack. will try to understand the root cause of the issue.

**Comment by kshtsk** [comment]
@tchaikov I spend some time for investigation, seems like this patch is really making a new connection all the time which is wrong because it removed the feature of reusing connection and doing the ensure_online always, which is wrong, example:
```
worker@teuth-kyr:~/src/teuthology_master> python
Python 3.6.10 (default, Jan 16 2020, 09:12:04) [GCC] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> from teuthology.orchestra.remote import Remote as r
>>> t=r('ubuntu@target-kyr-001.ovh.suse.de')
>>> t.sh('uptime')
2020-05-08 10:39:18,066.066 INFO:teuthology.orchestra.run.target-kyr-001:> true
2020-05-08 10:39:18,804.804 INFO:teuthology.orchestra.run.target-kyr-001:> uptime
2020-05-08 10:39:18,862.862 INFO:teuthology.orchestra.run.target-kyr-001.stdout: 10:39:18 up 2 min,  0 users,  load average: 0.15, 0.12, 0.04
' 10:39:18 up 2 min,  0 users,  load average: 0.15, 0.12, 0.04\n'
>>> t.sh('uptime')
2020-05-08 10:39:22,298.298 INFO:teuthology.orchestra.run.target-kyr-001:> true
2020-05-08 10:39:22,310.310 INFO:teuthology.orchestra.run.target-kyr-001:> uptime
2020-05-08 10:39:22,366.366 INFO:teuthology.orchestra.run.target-kyr-001.stdout: 10:39:21 up 2 min,  0 users,  load average: 0.13, 0.11, 0.04
' 10:39:21 up 2 min,  0 users,  load average: 0.13, 0.11, 0.04\n'
>>> 

```
As you can see the 'true' is running before each command, it should be only running once or if the connection gott lost.

**Comment by kshtsk** [comment]
Originally the usage should be looking like:
```
python
Python 3.6.10 (default, Jan 16 2020, 09:12:04) [GCC] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> from teuthology.orchestra.remote import Remote as r
>>> t=r('ubuntu@target-kyr-001.ovh.suse.de')
>>> t.sh('uptime')
2020-05-08 10:44:38,979.979 INFO:teuthology.orchestra.remote:Trying to reconnect to host
2020-05-08 10:44:39,090.090 INFO:teuthology.orchestra.run.target-kyr-001:> true
2020-05-08 10:44:39,874.874 INFO:teuthology.orchestra.run.target-kyr-001:> uptime
2020-05-08 10:44:39,925.925 INFO:teuthology.orchestra.run.target-kyr-001.stdout: 10:44:39 up 7 min,  0 users,  load average: 0.00, 0.03, 0.01
' 10:44:39 up 7 min,  0 users,  load average: 0.00, 0.03, 0.01\n'
>>> t.sh('uptime')
2020-05-08 10:44:43,069.069 INFO:teuthology.orchestra.run.target-kyr-001:> uptime
2020-05-08 10:44:43,080.080 INFO:teuthology.orchestra.run.target-kyr-001.stdout: 10:44:42 up 7 min,  0 users,  load average: 0.00, 0.03, 0.01
' 10:44:42 up 7 min,  0 users,  load average: 0.00, 0.03, 0.01\n'
>>> 
```

**Comment by tchaikov** [comment]
>  it removed the feature of reusing connection 

i don't think "true" is called using a new connection.

**Comment by kshtsk** [comment]
> > it removed the feature of reusing connection
> 
> i don't think "true" is called using a new connection.

@tchaikov the "true" is called because 'ensure_online' is added, which call 'is_online' in turn.

**Comment by kshtsk** [comment]
@tchaikov and probably you right, there might be no new connection, but the extra `run("true")` which does not make sense.

**Comment by tchaikov** [comment]
#1468

**Pull request merged**